sale_coupon_reward_add_product/models/sale_order.py

```diff
@@ -0,0 +1,37 @@
+# Copyright 2022 Camptocamp SA (https://www.camptocamp.com).
+# @author Iván Todorovich <ivan.todorovich@camptocamp.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _is_reward_in_order_lines(self, program):
+        if not self.env.context.get("check_reward_in_order_lines", True):
+            return True
+        return super()._is_reward_in_order_lines(program)
+
+    def _prepare_reward_product_line_vals(self, program):
+        return {
+            "order_id": self.id,
+            "product_id": program.reward_product_id.id,
+            "product_uom": program.reward_product_id.uom_id.id,
+        }
+
+    def _create_reward_line(self, program):
+        # OVERRIDE to add the reward product to the order automatically
+        if program.reward_product_add_if_missing:
+            reward_product_lines = self.order_line.filtered(
+                lambda line: line.product_id == program.reward_product_id
+            )
+            order_quantity = sum(reward_product_lines.mapped("product_uom_qty"))
+            if order_quantity < program.reward_product_quantity:
+                qty_to_add = float(program.reward_product_quantity) - order_quantity
+                line_vals = self._prepare_reward_product_line_vals(program)
+                line = self.env["sale.order.line"].new(line_vals)
+                line.product_uom_qty = qty_to_add
+                line.product_id_change()
```

Comment:
would be better to use onchange_helper to have all onchanges played properly

Reply:
Or `odoo.tests.Form` which doesn't pull extra dependencies

Reply:
So, I tested both approaches and it turns out `Form` is ~60% faster than `onchange_helper`. 
Not having to depend on an extra module is the cherry on top ❤️

Reply:
Many times debated. Form is not intended to be used outside tests... If onchange helper is slower than Form, I would be pleased to have metrics to fix it if needed but that is not the result we had

Reply:
@ivantodorovich could you provide some metrics? I'm curious :)

Reply:
> Many times debated. Form is not intended to be used outside tests

Makes sense, indeed.

> @ivantodorovich could you provide some metrics? I'm curious :)

So, I can't reach the same conclusion.
After measuring again it seems `onchange_helper` is actually faster. 🤔 
I don't remember what I did before, but it may be related to the project I was evaluating on, as I did find something curious

Take this example:

```py
class HelperTest(models.Model):
    _name = "helper.test"

    name = fields.Char()
    changed = fields.Integer()
    computed = fields.Integer(compute="_compute_computed")
    computed_group = fields.Integer(compute="_compute_computed_group", groups="base.group_portal")
    stored = fields.Integer(compute="_compute_stored", store=True)

    @api.onchange("name")
    def _onchange_name(self):
        print("_onchange_name")
        self.changed = len(self.name or "")
    
    @api.depends("name")
    def _compute_computed(self):
        print("_compute_computed")
        for rec in self:
            rec.computed = len(rec.name or "")

    @api.depends("name")
    def _compute_computed_group(self):
        print("_compute_computed_group")
        for rec in self:
            rec.computed_group = len(rec.name or "")
    
    @api.depends("name")
    def _compute_stored(self):
        print("_compute_stored")
        for rec in self:
            rec.stored = len(rec.name or "")

    @api.model
    def create_with_form(self):
        with Form(self) as form:
            form.name = "Testing"

    @api.model
    def create_with_helper(self):
        vals = {"name": "Testing"}
        vals.update(self.play_onchanges(vals, vals.keys()))
        self.create(vals)

```

```py
>>> self.env["helper.test"].create_with_form()
_onchange_name
_compute_computed
_compute_stored
_onchange_name
_compute_computed
_compute_stored
_compute_stored
_compute_computed

>>> self.env["helper.test"].create_with_helper()
_onchange_name
_compute_computed
_compute_computed_group
_compute_stored

```

- Both are running onchanges
- Both are computing non-stored computed fields (seems useless IMO, for our use case)
- Both are computing stored computed fields
- `Form` is less efficient, recomputing several times 🤔 
- `Form` is not computing fields it doesn't have access to, whilst `onchange_helper` is

Non-stored computed fields usually include fields like `_compute_helpdesk_ticket_count`, which will run sql queries to compute. Besides being useless to compute this, it may raise AccessError if the user doesn't have enough rights

So bottom line I think an improvement can be done to avoid computing non-stored fields altogether